File: algoritmos/morse.cs
using System;

public class Test {
    public static void Main() {
        decrypt("FENDSVTSLHW.EDATS,EULAY");
    }

    public static void decrypt(string entrada) {
        string caracteres = "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z _ , . ?";
        string[] morse = ".- -... -.-. -.. . ..-. --. .... .. .--- -.- .-.. -- -. --- .--. --.- .-. ... - ..- ...- .-- -..- -.-- --.. ..-- .-.- ---. ----".Split(" ");
        string numeros = "";
        string coded = "";

        for (int i = 0; i < entrada.Length; i++) {
            int indice = caracteres.IndexOf(entrada[i]) / 2;
            numeros += morse[indice].Length;
            coded += morse[indice];
        }

        string respuesta = "";


        for (int i = numeros.Length - 1; i >= 0; i--) {
            int j = Convert.ToInt32(Char.GetNumericValue(numeros[i])); //número de caracteres
            string subMorse = coded.Substring(0, j);
            coded = coded.Substring(j);
            int indice = Array.IndexOf(morse, subMorse);
            respuesta += caracteres[indice * 2];

        }
        Console.WriteLine(respuesta);
    }

}

File: algoritmos/morse.py
mensaje = "FENDSVTSLHW.EDATS,EULAY"
caracteres = "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z _ , . ?"
morse = ".- -... -.-. -.. . ..-. --. .... .. .--- -.- .-.. -- -. --- .--. --.- .-. ... - ..- ...- .-- -..- -.-- --.. ..-- .-.- ---. ----".split()
numeros = "";
coded = "";

for i in range(0, len(mensaje)):
    indice = int(caracteres.index(mensaje[i]) / 2)
    numeros += str(len(morse[indice]))
    coded += morse[indice]
    
respuesta = ""

i = len(numeros) - 1

while(i>=0):
    j = int(numeros[i])
    subMorse = coded[0: j]
    coded = coded[j:]
    indice = morse.index(subMorse)
    respuesta += caracteres[indice * 2]
    i-=1
    
print(respuesta)